crates/proposer-client/src/lib.rs

```diff
@@ -0,0 +1,92 @@
+use std::fmt::Display;
+use std::sync::Arc;
+use std::time::Duration;
+
+use serde::{Deserialize, Serialize};
+use sp1_sdk::SP1ProofWithPublicValues;
+
+pub use crate::config::ProposerClientConfig;
+pub use crate::error::Error;
+use crate::network_prover::AggProofProver;
+use crate::rpc::AggProofProposer;
+
+pub mod config;
+pub mod error;
+pub mod network_prover;
+pub mod rpc;
+
+/// The Proposer client is responsible for retrieval of the AggProof.
+/// AggProof is the aggregated proof of the multiple
+/// block span full execution proofs.
+///
+/// The proposer client communicates with the proposer API to
+/// request creation of the AggProof (getting the proof ID in return),
+/// and directly communicates with the SP1 cluster using NetworkProver
+/// to retrieve the generated proof.
+#[derive(Clone)]
+pub struct ProposerClient<AggProofProposerT, AggProofProverT>
+where
+    AggProofProposerT: AggProofProposer,
+    AggProofProverT: AggProofProver,
+{
+    proposer: Arc<AggProofProposerT>,
+    prover: Arc<AggProofProverT>,
+    proving_timeout: Option<Duration>,
+}
+
+impl<AggProofProposerT: AggProofProposer, AggProofProverT: AggProofProver>
+    ProposerClient<AggProofProposerT, AggProofProverT>
+{
+    pub fn new(
+        proposer: AggProofProposerT,
+        prover: AggProofProverT,
+        timeout: Option<Duration>,
+    ) -> Result<Self, error::Error> {
+        Ok(ProposerClient {
+            proposer: Arc::new(proposer),
+            prover: Arc::new(prover),
+            proving_timeout: timeout,
+        })
+    }
+
+    pub async fn request_agg_proof(&mut self, request: Request) -> Result<ProofId, Error> {
```

Comment:
You can bound the trait just where you need if I'm not wrong.

```suggestion
impl<Proposer, Prover>
    ProposerClient<Proposer, Prover>
{
    pub fn new(
        proposer: Arc<Proposer>,
        prover: Arc<Prover>,
        proving_timeout: Option<Duration>,
    ) -> Result<Self, error::Error> {
        Ok(ProposerClient { propose, prover, proving_timeout })
    }
}

impl<Proposer, Prover>
    ProposerClient<Proposer, Prover>
  where
  	Proposer: AggProofProposer,
  	Prover: AggProofProver
{
    pub async fn request_agg_proof(&mut self, request: Request) -> Result<ProofId, Error> {
```

Reply:
I can. But it doesn't make any real difference?